File: src/verify_8x8.c
/*
 * verify_8x8.c — Independent exhaustive enumeration of 8×8 Barron Squares.
 *
 * This is a VERIFICATION implementation, structurally independent of
 * barron8x8.c. It uses a different search decomposition to confirm
 * the count of 1,248 order-2 Barron Squares.
 *
 * Algorithm: Full Corner Derivation (FCD)
 *   Free variables: TL (4 digits), TR (4 digits), BL (4 digits)  [= 12 free]
 *   Derived:        BR from edge row/col constraints, inner block from consistency
 *
 * This differs from barron8x8.c's BOT-derivation, which uses:
 *   Free variables: TL (4), TR (4), col4_B (2), col5_B (2)       [= 12 free]
 *   Derived:        BL, BR, inner block via bottom-of-table chain
 *
 * The two implementations share no code and use different derivation
 * strategies, so agreement on the count constitutes independent verification.
 *
 * Compile:  gcc -O3 -o src/verify_8x8 src/verify_8x8.c
 * Run (single core, ~30h):
 *           ./src/verify_8x8
 * Run (12 parallel workers, ~3h wall):
 *           mkdir -p /tmp/verify_8x8
 *           for i in $(seq 0 11); do
 *             S=$((i*547)); E=$(((i+1)*547));
 *             [ $i -eq 11 ] && E=6561;
 *             ./src/verify_8x8 $S $E > /tmp/verify_8x8/out_$i.txt 2>/dev/null &
 *           done
 *           wait
 *           grep "^FOUND:" /tmp/verify_8x8/out_*.txt | \
 *             awk -F: '{s+=$NF} END {print "Total:", s}'
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

/* ------------------------------------------------------------------ */
/* Pre-computed tables                                                 */
/* ------------------------------------------------------------------ */

/*
 * edge_valid[L][R] = 1 iff L*R is in [1000,9999] with all 4 digits nonzero,
 *                    AND both L and R have nonzero digits.
 * edge_prod[L][R][0..3] = the 4 digits of L*R when edge_valid.
 */
static int edge_valid[100][100];
static int edge_prod[100][100][4];

/* For each L, sorted list of valid R values for edge constraints. */
static int edge_R_list[100][90];
static int edge_R_count[100];

static void precompute_tables(void) {
    memset(edge_valid, 0, sizeof(edge_valid));
    memset(edge_R_count, 0, sizeof(edge_R_count));

    for (int L = 11; L <= 99; L++) {
        if (L % 10 == 0) continue;
        for (int R = 11; R <= 99; R++) {
            if (R % 10 == 0) continue;
            int p = L * R;
            if (p < 1000 || p > 9999) continue;
            int d0 = p / 1000;
            int d1 = (p / 100) % 10;
            int d2 = (p / 10) % 10;
            int d3 = p % 10;
            if (d0 == 0 || d1 == 0 || d2 == 0 || d3 == 0) continue;

            edge_valid[L][R] = 1;
            edge_prod[L][R][0] = d0;
            edge_prod[L][R][1] = d1;
            edge_prod[L][R][2] = d2;
            edge_prod[L][R][3] = d3;
            edge_R_list[L][edge_R_count[L]++] = R;
        }
    }
}

/* ------------------------------------------------------------------ */
/* Inner-block consistency check with early exit                       */
/* ------------------------------------------------------------------ */

/*
 * Check that the 4×4 inner center derived from row products matches
 * the one derived from column products. Returns 1 if consistent.
 *
 * Row derivation: for inner row r, product = L_r * R_r → 4 digits
 * Col derivation: for inner col c, product = T_c * B_c → 4 digits
 * Cell (r,c) must agree: row_digit[r][c-2] == col_digit[c][r-2]
 *
 * Uses early-exit: check one product at a time, bail on first mismatch.
 */
static inline int check_inner_consistency(
    int L2, int R2, int L3, int R3, int L4, int R4, int L5, int R5,
    int T2, int B2, int T3, int B3, int T4, int B4, int T5, int B5)
{
    /* Compute all 8 products */
    int pr2 = L2 * R2, pr3 = L3 * R3, pr4 = L4 * R4, pr5 = L5 * R5;
    int pc2 = T2 * B2, pc3 = T3 * B3, pc4 = T4 * B4, pc5 = T5 * B5;

    if (pr2 > 9999 || pr3 > 9999 || pr4 > 9999 || pr5 > 9999) return 0;
    if (pc2 > 9999 || pc3 > 9999 || pc4 > 9999 || pc5 > 9999) return 0;

    /* Extract row digits: row_d[r][c] where r=0..3, c=0..3 */
    /* Row r gives digits of pr{r+2} */
    int rd[4][4];
    rd[0][0] = pr2/1000; rd[0][1] = (pr2/100)%10; rd[0][2] = (pr2/10)%10; rd[0][3] = pr2%10;
    rd[1][0] = pr3/1000; rd[1][1] = (pr3/100)%10; rd[1][2] = (pr3/10)%10; rd[1][3] = pr3%10;
    rd[2][0] = pr4/1000; rd[2][1] = (pr4/100)%10; rd[2][2] = (pr4/10)%10; rd[2][3] = pr4%10;
    rd[3][0] = pr5/1000; rd[3][1] = (pr5/100)%10; rd[3][2] = (pr5/10)%10; rd[3][3] = pr5%10;

    /* Check column by column with early exit */
    /* Col 2 (c=0): digits of pc2 should match rd[0..3][0] */
    if (rd[0][0] != pc2/1000) return 0;
    if (rd[1][0] != (pc2/100)%10) return 0;
    if (rd[2][0] != (pc2/10)%10) return 0;
    if (rd[3][0] != pc2%10) return 0;

    /* Col 3 */
    if (rd[0][1] != pc3/1000) return 0;
    if (rd[1][1] != (pc3/100)%10) return 0;
    if (rd[2][1] != (pc3/10)%10) return 0;
    if (rd[3][1] != pc3%10) return 0;

    /* Col 4 */
    if (rd[0][2] != pc4/1000) return 0;
    if (rd[1][2] != (pc4/100)%10) return 0;
    if (rd[2][2] != (pc4/10)%10) return 0;
    if (rd[3][2] != pc4%10) return 0;

    /* Col 5 */
    if (rd[0][3] != pc5/1000) return 0;
    if (rd[1][3] != (pc5/100)%10) return 0;
    if (rd[2][3] != (pc5/10)%10) return 0;
    if (rd[3][3] != pc5%10) return 0;

    return 1;
}

/* ------------------------------------------------------------------ */
/* Main search                                                         */
/* ------------------------------------------------------------------ */

int main(int argc, char **argv) {
    int tl_start = 0, tl_end = 6561;

    if (argc >= 3) {
        tl_start = atoi(argv[1]);
        tl_end   = atoi(argv[2]);
    }

    fprintf(stderr, "verify_8x8: Independent 8x8 Barron Square enumeration\n");
    fprintf(stderr, "Algorithm: Full Corner Derivation (TL -> TR -> BL -> derive BR)\n");
    fprintf(stderr, "TL range: [%d, %d) of 6561\n\n", tl_start, tl_end);

    precompute_tables();

    int total_edge_pairs = 0;
    for (int L = 11; L <= 99; L++)
        if (L % 10 != 0)
            total_edge_pairs += edge_R_count[L];
    fprintf(stderr, "Pre-computed %d valid edge (L,R) pairs.\n", total_edge_pairs);

    clock_t t_start_cpu = clock();
    time_t  t_start_wall = time(NULL);

    long long total_found = 0;
    int tl_idx = 0;

    for (int a = 1; a <= 9; a++)
    for (int b = 1; b <= 9; b++) {
        int L0 = 10*a + b;
        if (edge_R_count[L0] == 0) { tl_idx += 81; continue; }

        for (int c = 1; c <= 9; c++) {
            int T0 = 10*a + c;
            if (edge_R_count[T0] == 0) { tl_idx += 9; continue; }

            for (int d = 1; d <= 9; d++) {
                /* Skip TL values outside our assigned range */
                if (tl_idx < tl_start || tl_idx >= tl_end) {
                    tl_idx++;
                    continue;
                }
                tl_idx++;

                int L1 = 10*c + d;
                int T1 = 10*b + d;

                if (edge_R_count[L1] == 0) continue;
                if (edge_R_count[T1] == 0) continue;

                /* Iterate valid TR (from rows 0,1 edge constraints) */
                for (int ir0 = 0; ir0 < edge_R_count[L0]; ir0++) {
                    int R0 = edge_R_list[L0][ir0];
                    int e = R0 / 10, f = R0 % 10;
                    int *row0 = edge_prod[L0][R0];

                    for (int ir1 = 0; ir1 < edge_R_count[L1]; ir1++) {
                        int R1 = edge_R_list[L1][ir1];
                        int g = R1 / 10, h = R1 % 10;
                        int *row1 = edge_prod[L1][R1];

                        int T6 = 10*e + g;
                        int T7 = 10*f + h;

                        /* Skip if no valid bottom endpoints for cols 6 or 7 */
                        if (edge_R_count[T6] == 0 || edge_R_count[T7] == 0)
                            continue;

                        /* Inner col top endpoints (from row 0,1 interiors) */
                        int T2 = 10*row0[0] + row1[0];
                        int T3 = 10*row0[1] + row1[1];
                        int T4 = 10*row0[2] + row1[2];
                        int T5 = 10*row0[3] + row1[3];

                        /* Iterate valid BL (from cols 0,1 edge constraints) */
                        for (int ib0 = 0; ib0 < edge_R_count[T0]; ib0++) {
                            int B0 = edge_R_list[T0][ib0];
                            int *col0 = edge_prod[T0][B0];
                            int bi = B0 / 10, bk = B0 % 10;

                            for (int ib1 = 0; ib1 < edge_R_count[T1]; ib1++) {
                                int B1 = edge_R_list[T1][ib1];
                                int *col1 = edge_prod[T1][B1];
                                int bj = B1 / 10, bl_ = B1 % 10;

                                int left6 = 10*bi + bj;
                                int left7 = 10*bk + bl_;

                                if (edge_R_count[left6] == 0 || edge_R_count[left7] == 0)
                                    continue;

                                /* Inner row left endpoints (from col 0,1 interiors) */
                                int L2 = 10*col0[0] + col1[0];
                                int L3 = 10*col0[1] + col1[1];
                                int L4 = 10*col0[2] + col1[2];
                                int L5 = 10*col0[3] + col1[3];

                                /* Derive BR from 4 edge constraints */
                                for (int br00 = 1; br00 <= 9; br00++) {
                                    for (int br01 = 1; br01 <= 9; br01++) {
                                        int right6 = 10*br00 + br01;
                                        if (!edge_valid[left6][right6]) continue;

                                        for (int br10 = 1; br10 <= 9; br10++) {
                                            int bottom6 = 10*br00 + br10;
                                            if (!edge_valid[T6][bottom6]) continue;

                                            for (int br11 = 1; br11 <= 9; br11++) {
                                                int right7 = 10*br10 + br11;
                                                if (!edge_valid[left7][right7]) continue;

                                                int bottom7 = 10*br01 + br11;
                                                if (!edge_valid[T7][bottom7]) continue;

                                                /* BR passes all 4 edge checks.
                                                 * Derive inner endpoints and check consistency. */
                                                int *c6 = edge_prod[T6][bottom6];
                                                int *c7 = edge_prod[T7][bottom7];
                                                int *r6 = edge_prod[left6][right6];
                                                int *r7 = edge_prod[left7][right7];

                                                /* Check cols 6,7 interior digits nonzero (edge) */
                                                if (c6[0]==0||c7[0]==0||c6[1]==0||c7[1]==0||
                                                    c6[2]==0||c7[2]==0||c6[3]==0||c7[3]==0)
                                                    continue;
                                                /* Check rows 6,7 interior digits nonzero (edge) */
                                                if (r6[0]==0||r7[0]==0||r6[1]==0||r7[1]==0||
                                                    r6[2]==0||r7[2]==0||r6[3]==0||r7[3]==0)
                                                    continue;

                                                /* Inner row right endpoints */
                                                int R2 = 10*c6[0]+c7[0];
                                                int R3 = 10*c6[1]+c7[1];
                                                int R4 = 10*c6[2]+c7[2];
                                                int R5 = 10*c6[3]+c7[3];

                                                /* Inner col bottom endpoints */
                                                int B2 = 10*r6[0]+r7[0];
                                                int B3 = 10*r6[1]+r7[1];
                                                int B4 = 10*r6[2]+r7[2];
                                                int B5 = 10*r6[3]+r7[3];

                                                if (check_inner_consistency(
                                                        L2,R2,L3,R3,L4,R4,L5,R5,
                                                        T2,B2,T3,B3,T4,B4,T5,B5))
                                                {
                                                    total_found++;
                                                }
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                }

                /* Progress every 100 TL values */
                if ((tl_idx - tl_start) % 100 == 0 && tl_idx > tl_start) {
                    double elapsed = difftime(time(NULL), t_start_wall);
                    int done = tl_idx - tl_start;
                    int total = tl_end - tl_start;
                    double pct = 100.0 * done / total;
                    fprintf(stderr, "  TL %d/%d (%.1f%%), found %lld, %.0fs elapsed\n",
                            done, total, pct, total_found, elapsed);
                }
            }
        }
    }

    double cpu_secs = (double)(clock() - t_start_cpu) / CLOCKS_PER_SEC;
    double wall_secs = difftime(time(NULL), t_start_wall);

    printf("FOUND:%lld\n", total_found);
    fprintf(stderr, "\n========================================\n");
    fprintf(stderr, "TL range: [%d, %d)\n", tl_start, tl_end);
    fprintf(stderr, "Squares found: %lld\n", total_found);
    fprintf(stderr, "CPU time:  %.1f seconds (%.2f hours)\n", cpu_secs, cpu_secs/3600);
    fprintf(stderr, "Wall time: %.0f seconds (%.2f hours)\n", wall_secs, wall_secs/3600);
    fprintf(stderr, "========================================\n");

    return 0;
}

File: src/verify_8x8.py
#!/usr/bin/env python3
"""
verify_8x8.py — Independent verification of the 8x8 Barron Square count.

Two-phase verification:
  Phase 1: Validate all 1,248 stored squares using an independent checker
           (no shared code with barron8x8.c).
  Phase 2: Fresh exhaustive enumeration via verify_8x8.c (Full Corner
           Derivation algorithm), which uses a fundamentally different
           search decomposition from barron8x8.c's BOT-derivation.

Phase 2 details (verify_8x8.c):
  Algorithm: Full Corner Derivation (FCD)
    Free variables: TL (4 digits), TR (4 digits), BL (4 digits)
    Derived: BR from edge row/col constraints, inner block from consistency
  This differs from barron8x8.c's BOT-derivation:
    Free variables: TL (4), TR (4), col4_B (2), col5_B (2)
    Derived: BL, BR, inner block via bottom-of-table chain

  Verification result (2026-04-14/15):
    12 parallel workers, 160.8 total CPU hours, 19.2h max wall time.
    All 12 per-worker counts match expected, total = 1,248. PASSED.

Usage:
    # Phase 1 only (validation, seconds):
    python3 src/verify_8x8.py --validate-only

    # Phase 2: compile and run verify_8x8.c (see src/verify_8x8.c header)

    # Collect Phase 2 results:
    python3 src/collect_verify.py /tmp/verify_8x8/out_*.txt
"""

import json
import sys
import os
import time
import argparse


# ---------------------------------------------------------------------------
# Phase 1: Independent validation of stored squares
# ---------------------------------------------------------------------------

def validate_barron_8x8(M):
    """
    Validate that M is an 8x8 Barron Square (order k=2).

    Checks:
      - 8x8 dimensions
      - All 8 row constraints: left(2-digit) * right(2-digit) = interior(4-digit)
      - All 8 column constraints: top(2-digit) * bottom(2-digit) = interior(4-digit)
      - No zeros in edge cells (rows 0,1,6,7 or cols 0,1,6,7)

    This is a standalone implementation with no imports from barron_core.py.
    """
    if len(M) != 8 or any(len(row) != 8 for row in M):
        return False, "Wrong dimensions"

    # Check digit range
    for i in range(8):
        for j in range(8):
            if not (0 <= M[i][j] <= 9):
                return False, f"Cell ({i},{j}) = {M[i][j]} out of range"

    # Check no zeros on edges
    for i in range(8):
        for j in range(8):
            is_edge = (i < 2 or i >= 6 or j < 2 or j >= 6)
            if is_edge and M[i][j] == 0:
                return False, f"Zero at edge cell ({i},{j})"

    # Check row constraints
    for i in range(8):
        left = 10 * M[i][0] + M[i][1]
        right = 10 * M[i][6] + M[i][7]
        interior = 1000 * M[i][2] + 100 * M[i][3] + 10 * M[i][4] + M[i][5]
        if left * right != interior:
            return False, f"Row {i}: {left} * {right} = {left*right} != {interior}"

    # Check column constraints
    for j in range(8):
        top = 10 * M[0][j] + M[1][j]
        bottom = 10 * M[6][j] + M[7][j]
        interior = 1000 * M[2][j] + 100 * M[3][j] + 10 * M[4][j] + M[5][j]
        if top * bottom != interior:
            return False, f"Col {j}: {top} * {bottom} = {top*bottom} != {interior}"

    return True, "Valid"


def phase1_validate(squares_path):
    """Validate all stored 8x8 squares."""
    print("=" * 60)
    print("PHASE 1: Validating stored 8x8 Barron Squares")
    print("=" * 60)

    with open(squares_path) as f:
        squares = json.load(f)

    print(f"Loaded {len(squares)} squares from {squares_path}")

    t0 = time.time()
    failures = []
    for idx, sq in enumerate(squares):
        ok, msg = validate_barron_8x8(sq)
        if not ok:
            failures.append((idx, msg))

    elapsed = time.time() - t0

    if failures:
        print(f"\nFAILED: {len(failures)} squares failed validation:")
        for idx, msg in failures[:10]:
            print(f"  Square #{idx}: {msg}")
        return False
    else:
        print(f"\nPASSED: All {len(squares)} squares are valid 8x8 Barron Squares")
        print(f"Validation time: {elapsed:.3f}s")

        # Check for duplicates
        tuples = set()
        for sq in squares:
            tuples.add(tuple(tuple(row) for row in sq))
        if len(tuples) != len(squares):
            print(f"WARNING: {len(squares) - len(tuples)} duplicate squares found!")
            return False
        else:
            print(f"No duplicates: {len(tuples)} unique squares confirmed")
        return True


# ---------------------------------------------------------------------------
# Phase 2: Z3-based exhaustive enumeration
# ---------------------------------------------------------------------------

def phase2_enumerate(expected_count=1248):
    """
    Enumerate all 8x8 Barron Squares using Z3 bitvector SMT solver.

    Encodes each of the 64 cells as a BitVec(8) variable with:
      - Domain constraints: 0-9 for inner cells, 1-9 for edge cells
      - 8 row constraints: left * right = interior
      - 8 column constraints: top * bottom = interior

    Uses iterative blocking on the 16 corner variables to enumerate
    all solutions without redundancy.
    """
    print("\n" + "=" * 60)
    print("PHASE 2: Z3 exhaustive enumeration of 8x8 Barron Squares")
    print("=" * 60)

    try:
        from z3 import BitVec, BitVecVal, Solver, sat, Or, ZeroExt, UGE, ULE
    except ImportError:
        print("ERROR: z3-solver not available. Install with: pip install z3-solver")
        return None

    print("Building Z3 model...")
    t0 = time.time()

    s = Solver()

    # Create 64 cell variables (8-bit bitvectors, range 0-255)
    M = [[BitVec(f'm{i}{j}', 8) for j in range(8)] for i in range(8)]

    # Domain constraints
    for i in range(8):
        for j in range(8):
            s.add(ULE(M[i][j], BitVecVal(9, 8)))
            is_edge = (i < 2 or i >= 6 or j < 2 or j >= 6)
            if is_edge:
                s.add(UGE(M[i][j], BitVecVal(1, 8)))

    # Helper: zero-extend to 16-bit for multiplication
    def ext(v):
        return ZeroExt(8, v)

    ten = BitVecVal(10, 16)
    hundred = BitVecVal(100, 16)
    thousand = BitVecVal(1000, 16)

    # Row constraints: left * right = interior
    for i in range(8):
        left = ext(M[i][0]) * ten + ext(M[i][1])
        right = ext(M[i][6]) * ten + ext(M[i][7])
        interior = (ext(M[i][2]) * thousand + ext(M[i][3]) * hundred +
                    ext(M[i][4]) * ten + ext(M[i][5]))
        s.add(left * right == interior)

    # Column constraints: top * bottom = interior
    for j in range(8):
        top = ext(M[0][j]) * ten + ext(M[1][j])
        bottom = ext(M[6][j]) * ten + ext(M[7][j])
        interior = (ext(M[2][j]) * thousand + ext(M[3][j]) * hundred +
                    ext(M[4][j]) * ten + ext(M[5][j]))
        s.add(top * bottom == interior)

    # Corner variables (for blocking clauses)
    corners = []
    for i in (0, 1, 6, 7):
        for j in (0, 1, 6, 7):
            corners.append(M[i][j])

    build_time = time.time() - t0
    print(f"Model built in {build_time:.2f}s ({len(corners)} corner variables)")
    print(f"Enumerating solutions (blocking on {len(corners)} corner vars)...")
    print()

    t_enum = time.time()
    count = 0
    solutions = []

    while True:
        result = s.check()
        if result != sat:
            break

        model = s.model()
        count += 1

        # Extract the matrix
        matrix = []
        for i in range(8):
            row = []
            for j in range(8):
                val = model.eval(M[i][j]).as_long()
                row.append(val)
            matrix.append(row)
        solutions.append(matrix)

        # Verify this solution with our independent checker
        ok, msg = validate_barron_8x8(matrix)
        if not ok:
            print(f"  ERROR: Z3 solution #{count} fails validation: {msg}")
            print(f"  Matrix: {matrix}")
            return None

        # Block this solution (on corner variables only)
        block = Or([c != model.eval(c) for c in corners])
        s.add(block)

        # Progress reporting
        if count % 50 == 0:
            elapsed = time.time() - t_enum
            rate = count / elapsed if elapsed > 0 else 0
            print(f"  Found {count:,} squares ({rate:.1f}/s, {elapsed:.1f}s elapsed)")

    total_time = time.time() - t_enum

    print(f"\nEnumeration complete.")
    print(f"  Total squares found: {count:,}")
    print(f"  Enumeration time: {total_time:.1f}s ({total_time/60:.1f} min)")
    print(f"  Average rate: {count/total_time:.2f} squares/s" if total_time > 0 else "")

    if count == expected_count:
        print(f"\n  PASSED: Count matches expected {expected_count}")
    else:
        print(f"\n  FAILED: Found {count}, expected {expected_count}")

    return solutions


# ---------------------------------------------------------------------------
# Phase 3 (optional): Cross-check Z3 solutions against stored squares
# ---------------------------------------------------------------------------

def cross_check(z3_solutions, stored_path):
    """Verify Z3 solutions match stored squares exactly."""
    print("\n" + "=" * 60)
    print("CROSS-CHECK: Comparing Z3 solutions with stored squares")
    print("=" * 60)

    with open(stored_path) as f:
        stored = json.load(f)

    z3_set = set(tuple(tuple(r) for r in sq) for sq in z3_solutions)
    stored_set = set(tuple(tuple(r) for r in sq) for sq in stored)

    only_z3 = z3_set - stored_set
    only_stored = stored_set - z3_set
    common = z3_set & stored_set

    print(f"  Z3 found:     {len(z3_set)} unique squares")
    print(f"  Stored:       {len(stored_set)} unique squares")
    print(f"  In common:    {len(common)}")
    print(f"  Only in Z3:   {len(only_z3)}")
    print(f"  Only stored:  {len(only_stored)}")

    if only_z3:
        print("\n  Squares found by Z3 but not in stored file:")
        for sq in list(only_z3)[:3]:
            print(f"    {list(list(r) for r in sq)}")

    if only_stored:
        print("\n  Squares in stored file but not found by Z3:")
        for sq in list(only_stored)[:3]:
            print(f"    {list(list(r) for r in sq)}")

    if not only_z3 and not only_stored:
        print("\n  PASSED: Perfect match between Z3 enumeration and stored squares")
        return True
    else:
        print("\n  FAILED: Mismatch detected")
        return False


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main():
    parser = argparse.ArgumentParser(description="Independent verification of 8x8 Barron Squares")
    parser.add_argument("--validate-only", action="store_true",
                        help="Only run Phase 1 (validate stored squares)")
    parser.add_argument("--enumerate-only", action="store_true",
                        help="Only run Phase 2 (Z3 enumeration)")
    parser.add_argument("--squares", default=None,
                        help="Path to 8x8_squares.json")
    args = parser.parse_args()

    # Find the squares file
    squares_path = args.squares
    if squares_path is None:
        script_dir = os.path.dirname(os.path.abspath(__file__))
        squares_path = os.path.join(script_dir, "..", "results", "8x8_squares.json")

    t_start = time.time()
    all_passed = True

    if not args.enumerate_only:
        if not os.path.exists(squares_path):
            print(f"ERROR: Squares file not found: {squares_path}")
            sys.exit(1)
        passed = phase1_validate(squares_path)
        all_passed = all_passed and passed

    z3_solutions = None
    if not args.validate_only:
        z3_solutions = phase2_enumerate(expected_count=1248)
        if z3_solutions is None:
            all_passed = False
        elif len(z3_solutions) != 1248:
            all_passed = False

    if z3_solutions and not args.enumerate_only and os.path.exists(squares_path):
        passed = cross_check(z3_solutions, squares_path)
        all_passed = all_passed and passed

    total_time = time.time() - t_start
    print(f"\n{'='*60}")
    print(f"VERIFICATION {'PASSED' if all_passed else 'FAILED'}")
    print(f"Total time: {total_time:.1f}s ({total_time/3600:.2f} hours)")
    print(f"{'='*60}")

    sys.exit(0 if all_passed else 1)


if __name__ == "__main__":
    main()
